deps/chakrashim/src/v8debug.cc

```diff
@@ -21,71 +21,715 @@
+      this->processShouldContinueFn, args, _countof(args), &result);
+    CHAKRA_VERIFY_NOERROR(errorCode);
+
+    JsValueType resultType;
+    errorCode = JsGetValueType(result, &resultType);
+    CHAKRA_VERIFY_NOERROR(errorCode);
+
+    CHAKRA_VERIFY(resultType == JsBoolean);
+
+    bool shouldContinue = true;
+    errorCode = JsBooleanToBool(result, &shouldContinue);
+    CHAKRA_VERIFY_NOERROR(errorCode);
+
+    return shouldContinue;
+  }
+
+  MessageImpl::MessageImpl(Local<String> v8Str) {
+    this->v8Str = v8Str;
+  }
+
+  Isolate* MessageImpl::GetIsolate() const {
+    return jsrt::IsolateShim::GetCurrentAsIsolate();
+  }
+
+  Local<String> MessageImpl::JsValueRefToV8String(JsValueRef result) {
+    JsValueRef scriptRef;
+    const wchar_t *script;
+
+    JsErrorCode errorCode = jsrt::ToString(result, &scriptRef, &script);
+    CHAKRA_VERIFY_NOERROR(errorCode);
+
+    v8::Local<v8::String> v8Str = v8::String::NewFromTwoByte(
+      jsrt::IsolateShim::GetCurrentAsIsolate(),
+      reinterpret_cast<uint16_t*>(const_cast<wchar_t*>(script)));
+    return v8Str;
+  }
+
+  Local<String> MessageImpl::GetJSON() const {
+    return this->v8Str;
+  }
+
+  void Debug::EnableDebug() {
     g_EnableDebug = true;
+  }
+
+  bool Debug::IsDebugEnabled() {
+    return g_EnableDebug;
+  }
+
+  Local<Context> Debug::GetDebugContext(Isolate* isolate) {
+    jsrt::IsolateShim* isoShim = jsrt::IsolateShim::FromIsolate(isolate);
+    JsContextRef debugContextRef = JS_INVALID_REFERENCE;
+    if (isoShim->debugContext == JS_INVALID_REFERENCE) {
+      HandleScope scope(isolate);
+      debugContextRef = *Context::New(isolate);
+      isoShim->debugContext = isoShim->GetContextShim(
+        debugContextRef);
+      JsAddRef(debugContextRef, nullptr);
+
+      Local<Object> global = Context::GetCurrent()->Global();
+
+      // CHAKRA-TODO: Chakra doesn't fully implement the debugger without
+      // --debug flag. Add a dummy 'Debug' on global object if it doesn't
+      // already exist.
+      {
+        Context::Scope context_scope(debugContextRef);
+        jsrt::ContextShim* contextShim = jsrt::ContextShim::GetCurrent();
+        JsValueRef ensureDebug = contextShim->GetensureDebugFunction();
+        JsValueRef unused;
+        if (jsrt::CallFunction(ensureDebug, *global, &unused) != JsNoError) {
+          return Local<Context>();
+        }
+      }
+    }
+
+    return static_cast<Context*>(isoShim->debugContext->GetContextRef());
+  }
 
-    Local<Context> currentContext = Context::GetCurrent();
-    if (!currentContext.IsEmpty()) {
-      // Turn on debug mode on current Context (script engine), which was
-      // created before start debugging and not in debug mode.
-      JsStartDebugging();
+  void CALLBACK JsDiagDebugEventHandler(
+    JsDiagDebugEvent debugEvent,
+    JsValueRef eventData,
+    void* callbackState) {
+    if (Debug::handler == nullptr) {
+      return;
     }
+
+    Debug::messageQueue.debugEventProcessCount++;
```

Comment:
JsDiagDebugEventHandler is supposed to be per runtime, so it is only called from one runtime at a time. No need for interlock.

---
In reply to: [81492243](https://github.com/nodejs/node-chakracore/pull/126#discussion_r81492243) [](ancestors = 81492243)